pkg/kv/kv.go

```diff
@@ -32,6 +35,141 @@ import (
 
 var extraHandleColumnInfo = model.NewExtraHandleColInfo()
 
+// Iter abstract iterator method for Ingester.
+type Iter interface {
+	// Seek seek to specify position.
+	// if key not found, seeks next key position in iter.
+	Seek(key []byte) bool
+	// Error return current error on this iter.
+	Error() error
+	// First return the first key in this iter.
+	First() bool
+	// Last return the last key in this iter.
+	Last() bool
+	// Valid check this iter reach the end.
+	Valid() bool
+	// Next moves this iter forward.
+	Next() bool
+	// Key represents current position pair's key.
+	Key() []byte
+	// Value represents current position pair's Value.
+	Value() []byte
+	// Close close this iter.
+	Close() error
+}
+
+// IterProducer produces iterator with given range.
+type IterProducer interface {
+	// Produce produces iterator with given range [start, end).
+	Produce(start []byte, end []byte) Iter
+}
+
+// SimpleKVIterProducer represents kv iter producer.
+type SimpleKVIterProducer struct {
+	pairs Pairs
+}
+
+// NewSimpleKVIterProducer creates SimpleKVIterProducer.
+func NewSimpleKVIterProducer(pairs Pairs) IterProducer {
+	return &SimpleKVIterProducer{
+		pairs: pairs,
+	}
+}
+
+// Produce implements Iter.Producer.Produce.
+func (p *SimpleKVIterProducer) Produce(start []byte, end []byte) Iter {
+	startIndex := sort.Search(len(p.pairs), func(i int) bool {
+		return bytes.Compare(start, p.pairs[i].Key) < 1
+	})
+	endIndex := sort.Search(len(p.pairs), func(i int) bool {
+		return bytes.Compare(end, p.pairs[i].Key) < 1
+	})
+	if startIndex >= endIndex {
+		log.Warn("produce failed due to start key is large than end key",
+			zap.Binary("start", start), zap.Binary("end", end))
+		return nil
+	}
+	return newSimpleKVIter(p.pairs[startIndex:endIndex])
+}
+
+// SimpleKVIter represents simple pair iterator.
+// which is used for log restore.
+type SimpleKVIter struct {
+	index int
+	pairs Pairs
+}
+
+// newSimpleKVIter creates SimpleKVIter.
+func newSimpleKVIter(pairs Pairs) Iter {
+	return &SimpleKVIter{
+		index: -1,
+		pairs: pairs,
+	}
+}
+
+// Seek implements Iter.Seek.
+func (s *SimpleKVIter) Seek(key []byte) bool {
+	s.index = sort.Search(len(s.pairs), func(i int) bool {
+		return bytes.Compare(key, s.pairs[i].Key) < 1
+	})
+	return s.index < len(s.pairs)
+}
+
+// Error implements Iter.Error.
+func (s *SimpleKVIter) Error() error {
+	return nil
+}
+
+// First implements Iter.First.
+func (s *SimpleKVIter) First() bool {
+	if len(s.pairs) == 0 {
+		return false
+	}
+	s.index = 0
+	return true
+}
+
+// Last implements Iter.Last.
+func (s *SimpleKVIter) Last() bool {
+	if len(s.pairs) == 0 {
+		return false
+	}
+	s.index = len(s.pairs) - 1
+	return true
+}
+
+// Valid implements Iter.Valid.
+func (s *SimpleKVIter) Valid() bool {
+	return s.index >= 0 && s.index < len(s.pairs)
+}
+
+// Next implements Iter.Next.
+func (s *SimpleKVIter) Next() bool {
+	s.index++
+	return s.index <= len(s.pairs)
```

Comment:
why use `<=` instead of `<` here 🤔 what does the boolean return supposed to mean

Reply:
fixed.`<=` is meaningless